library/cartridge_check_instance_started.py

```diff
@@ -4,25 +4,77 @@
 }
 
 
 def check_instance_started(params):
     try:
         control_console = get_control_console(params['control_sock'])
-        instance_is_alive = True
-        if not params['stateboard']:
-            instance_is_alive, _ = control_console.eval_res_err('''
-                return require('membership').myself().status == 'alive'
+
+        if params['stateboard']:
+            box_status, err = control_console.eval_res_err('''
+                if type(box.cfg) == 'function' then
+                    return nil, "Box hasn't been configured!"
+                end
+                return box.info.status
             ''')
+            if not box_status:
+                return ModuleRes(success=False, msg="Impossible to get stateboard status: %s!" % err)
+            if box_status != 'running':
+                return ModuleRes(success=False, msg="Stateboard is not running, it has '%s' status!" % box_status)
+
+        else:
+            member_status, err = control_console.eval_res_err("return require('membership').myself().status")
+            if not member_status:
+                return ModuleRes(success=False, msg="Impossible to get membership status: %s!" % err)
+            if member_status != 'alive':
+                return ModuleRes(success=False, msg="Membership is not alive, it has '%s' status!" % member_status)
+
+            cartridge_state, err = control_console.eval_res_err("return require('membership').myself().payload.state")
+            if not cartridge_state:
+                return ModuleRes(success=False, msg="Impossible to get instance Cartridge state: %s!" % err)
+            if cartridge_state not in params['expected_state']:
+                return ModuleRes(
+                    success=False,
+                    msg="Instance is not in one of states: %s, it's in '%s' state!" % (
+                        params['expected_state'],
+                        cartridge_state,
+                    ),
+                )
+
+            if params['wait_buckets_discovery']:
+                buckets_discovered, err = control_console.eval_res_err('''
+                    local router = require('vshard.router')
+                    if router.info == nil then
+                        return true
+                    end
+
+                    if require('cartridge.vshard-utils').can_bootstrap() then
+                        return true
+                    end
+
+                    if router.info().bucket.unknown == 0 then
+                        return true
+                    end
+
+                    return nil, string.format(
+                        '%s out of %s buckets are not discovered!',
+                        router.info().bucket.unknown,
+                        router.bucket_count()
+                    )
+                ''')
+                if not buckets_discovered:
+                    return ModuleRes(success=False, msg=err)
+
     except CartridgeException as e:
```

Comment:
CartridgeException is caught in `main` function, why do you catch it here?

Reply:
I didn't change this code, but I'll remove it